cpp/src/arrow/compute/memory_resources.cc

```diff
@@ -0,0 +1,307 @@
+#include "arrow/util/windows_compatibility.h"
+
+namespace arrow {
+
+namespace compute {
+
+std::string MemoryLevelName(MemoryLevel memory_level) {
+  static const char* MemoryLevelNames[] = {ARROW_STRINGIFY(MemoryLevel::kDiskLevel),
+                                           ARROW_STRINGIFY(MemoryLevel::kCPULevel),
+                                           ARROW_STRINGIFY(MemoryLevel::kGPULevel)};
+
+  return MemoryLevelNames[static_cast<int>(memory_level)];
+}
+
+std::string MemoryResource::ToString() const { return MemoryLevelName(memory_level_); }
+
+class CPUDataHolder : public DataHolder {
+ public:
+  explicit CPUDataHolder(const std::shared_ptr<RecordBatch>& record_batch)
+      : DataHolder(MemoryLevel::kCPULevel), record_batch_(std::move(record_batch)) {}
+
+  Result<ExecBatch> Get() override { return ExecBatch(*record_batch_); }
+
+ private:
+  std::shared_ptr<RecordBatch> record_batch_;
+};
+
+namespace {
+
+std::string RandomString(std::size_t length) {
+  const std::string characters =
+      "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+  std::random_device random_device;
+  std::mt19937 generator(random_device());
+  std::uniform_int_distribution<> distribution(0, characters.size() - 1);
+  std::string random_string;
+  for (std::size_t i = 0; i < length; ++i) {
+    random_string += characters[distribution(generator)];
+  }
+  return random_string;
+}
+
+}  // namespace
+
+Status StoreRecordBatch(const std::shared_ptr<RecordBatch>& record_batch,
+                        const std::shared_ptr<fs::FileSystem>& filesystem,
+                        const std::string& file_path) {
+  auto output = filesystem->OpenOutputStream(file_path).ValueOrDie();
+  auto writer =
+      arrow::ipc::MakeFileWriter(output.get(), record_batch->schema()).ValueOrDie();
+  ARROW_RETURN_NOT_OK(writer->WriteRecordBatch(*record_batch));
+  return writer->Close();
+}
+Result<std::shared_ptr<RecordBatch>> RecoverRecordBatch(
+    const std::shared_ptr<fs::FileSystem>& filesystem, const std::string& file_path) {
+  ARROW_ASSIGN_OR_RAISE(auto input, filesystem->OpenInputFile(file_path));
+  ARROW_ASSIGN_OR_RAISE(auto reader, arrow::ipc::feather::Reader::Open(input));
+  std::shared_ptr<Table> table;
+  ARROW_RETURN_NOT_OK(reader->Read(&table));
+  TableBatchReader batch_iter(*table);
+  ARROW_ASSIGN_OR_RAISE(auto batch, batch_iter.Next());
+  return batch;
+}
+
+class DiskDataHolder : public DataHolder {
+ public:
+  DiskDataHolder(const std::shared_ptr<RecordBatch>& record_batch,
+                 MemoryPool* memory_pool)
+      : DataHolder(MemoryLevel::kDiskLevel), memory_pool_(memory_pool) {
+    std::string root_path;
+    std::string file_name = "data-holder-temp-" + RandomString(64) + ".feather";
+
+    filesystem_ =
+        arrow::fs::FileSystemFromUri(cache_storage_root_path, &root_path).ValueOrDie();
+
+    file_path_ = root_path + file_name;
+    status_ = StoreRecordBatch(record_batch, filesystem_, file_path_);
```

Comment:
Complex logic shouldn't happen in constructors.  Maybe there needs to be a `Put` method or something.

Reply:
So there are a few options here but what we are saying is that we have these different DataHolders that are created from record_batch. We could make a function which is like

`template <typename DataHolderType> make_data_holder_from_batch(...)`

And it would invoke the constructur with something like the filesystem and path pointing to a file that has already persisted the record batch. This would allow the constructor to be simpler while still being able to have a common api for creating the data_holders from a record_batch